### PR TITLE
fix hostport checking for initContainers since they run in sequential order

### LINKED ISSUE #57443: make spec.HostPorts in initContainer clear
I found that kubelet will not allocate hostports specified from initContainers.
And PodFitsHostPorts in kube-scheduler does not take initContainer into consideration.

Also, kube-apiserver think such pod is illegal, but I think it OK.
```
$ cat two-files-counter-pod-streaming-sidecar.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: counter
spec:
  hostname: name
  subdomain: subdomain
  initContainers:
  - name: init-step1
    image: busybox
    args:
    - /bin/sh
    - -c
    - >
      sleep 10;
    ports:
      - containerPort: 8081
        hostPort: 8081
  - name: init-step2
    image: busybox
    args:
    - /bin/sh
    - -c
    - >
      sleep 10;
    ports:
      - containerPort: 8081
        hostPort: 8081    # this port does not conflict with the previous one, because initContainer runs in sequential order.
  containers:
  containers:
  - name: count
    image: busybox
    args:
    - /bin/sh
    - -c
    - >
      i=0;
      while true;
      do
        i=$((i+1));
        sleep $i;
      done
$ kubectl create -f two-files-counter-pod-streaming-sidecar.yaml 
The Pod "counter" is invalid: spec.initContainers[1].ports[0].hostPort: Duplicate value: "TCP//8081"
```

We should make the behavior of spec.HostPorts in initContainer clear

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
Fix hostport checking for initContainers since they run in sequential order

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
partial Fixes #57443

**Special notes for your reviewer**:
/assign @CaoShuFeng @dims 

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by dixudx** [comment]
ping @liggitt @smarterclayton for approval. Thanks.

**Comment by dchen1107** [comment]
Looks good to me, but leave @erictune for the final approval.

**Comment by dixudx** [comment]
ping @erictune for review. PTAL. Thanks.
/cc @erictune

**Comment by dixudx** [comment]
ping @erictune for review. PTAL. Thanks.

**Comment by dixudx** [comment]
Kindly ping @liggitt @smarterclayton for review. Thanks.

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -2520,8 +2520,13 @@ func validateContainers(containers []core.Container, volumes map[string]core.Vol
 		allErrs = append(allErrs, ValidateResourceRequirements(&ctr.Resources, idxPath.Child("resources"))...)
 		allErrs = append(allErrs, ValidateSecurityContext(ctr.SecurityContext, idxPath.Child("securityContext"))...)
 	}
-	// Check for colliding ports across all containers.
-	allErrs = append(allErrs, checkHostPortConflicts(containers, fldPath)...)
+
+	// HostPorts will not be in conflict since initContainers must
```
[the remark this answers is not in the document]
Comment (dixudx) [reply pkg/apis/core/validation/validation.go]: @CaoShuFeng Good catch. Thanks.

Updated. PTAL.

**Comment by dixudx** [comment]
@liggitt @smarterclayton PTAL. Thanks.

**Pull request merged**